docs/src/main/docs/health/02_health_in_k8s.adoc

```diff
@@ -0,0 +1,217 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Kubernetes probes
+:description: Kubernetes probes
+:keywords: helidon, readiness, liveliness, probes, kubernetes
+:kubernetes-probes-url: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes
+
+This document describes usages of the Helidon health check API with Kubernetes.
+
+== About Kubernetes probes
+
+Probes is the term used by Kubernetes to describe health checks for containers
+ (more info link:{kubernetes-probes-url}[here]).
+
+There are 2 types of probes:
+
+* `liveness`: Indicates whether the container is running
+* `readiness`: Indicates whether the container is ready to service requests
+
+Probes can be implemented with 3 different mechanisms:
+
+. Running a command inside a container
+. Making an `HTTP` request against a container
+. Opening a `TCP` socket against a container
+
+A microservice exposed to HTTP will typically implement both the liveness probe
+ and the readiness probe using HTTP requests.
+
+Probes can be configured with different parameters, see the most relevant
+ ones below:
+
+[cols="2,5",role="flex, sm7"]
+|=======
+| `initialDelaySeconds`
+| Number of seconds before probes are initiated
+
+| `periodSeconds`
+| How often to perform the probe. Default to 10 seconds. Minimum value is 1
+
+| `timeoutSeconds`
+| Number of seconds after which the probe times out. Defaults to 1 second.
+ Minimum value is 1
+
+| `failureThreshold`
+| Number of failed probes before giving up. Defaults to 3. Minimum value is 1.
+|=======
+
+=== Liveness probe
+
+The liveness probe is used to verify the container has become unresponsive,
+ for example it can detect deadlocks or analyze heap usage. When Kubernetes
+ gives up on a liveness probe, the corresponding pod is restarted.
+
+NOTE: A bad liveness probe can result in repeated restarts.
+
+We recommend the following:
+
+* Avoid checking dependencies in a liveness probe
+* Set `timeoutSeconds` to avoid excessive probe failures
+* Acknowledge startup times with `initDelaysSeconds`
+
+=== Readiness probe
+
+The readiness probe is used to avoid routing requests to the pod until it is
+ ready to accept traffic. When Kubernetes gives up on a readiness probe, the
+ pod is not restarted, it is removed from the service routing.
+
+NOTE: A bad readiness probe can result in a cascading effect where all pods are
+ removed from the service routing.
+
+We recommend the following:
+
+* Be conservative when checking shared dependencies
+* Be aggressive when checking local dependencies
+* Set `failureThreshold` according to `periodSeconds` in order to accomodate
+ temporary errors
+
+== Troubleshooting probes
```

Comment:
This is back again to general K8s information.

Reply:
Not sure what "again" refers to, the content before this is also general K8s information. Did  I miss a review comment ?

Reply:
I had typed -- but I guess I did not save -- a comment. Much of this document -- up through line 86 and then again lines 94-110 -- describes very useful information but it's about Kubernetes health in general, not anything specific about Helidon and K8s together. The "We recommend" part is from us at least even though it also applies to K8s in general. 

The example starting at 112 is, of course, Helidon-specific. 

I remember that we decided to streamline other parts of the doc by removing discussions of maven topics that were useful but not specific to Helidon. We decided we could assume that either the reader knew that maven info or would find it elsewhere.

Similarly, someone coming to this page who already knows about K8s health will not want to have to scroll past the general K8s health info to get to the Helidon-specific parts, and for readers who don't have the K8s background the link to K8s probes info might be sufficient.

If we remove the general information then the page is pretty short. That might be just fine -- get the reader the Helidon-relevant information as quickly and cleanly as possible. Or maybe there are some other things we could say about Helidon and K8s health on this page if we wanted to lengthen it.

Reply:
The point of having the general info is to make recommendations on how to best implement probes using the Helidon API.
Without the general information, it's hard to make such statements since they are about specific configuration properties.

See https://blog.colinbreck.com/kubernetes-liveness-and-readiness-probes-how-to-avoid-shooting-yourself-in-the-foot/

Reply:
To your point about "streamline the docs", I feel that Maven is common knowledge for a Java developer in 2019, however I do feel that we are all K8s newbies and we need to provide cheat sheets for certain topics.